core/extension/gdnative_interface.h

```diff
@@ -0,0 +1,416 @@
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef GDNATIVE_INTERFACE_H
+#define GDNATIVE_INTERFACE_H
+
+/* This is a C class header, you can copy it and use it directly in your own binders.
+ * Together with the JSON file, you should be able to generate any binder.
+ */
+
+#include <stdint.h>
+#include <stdio.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* VARIANT TYPES */
+
+enum GDNativeVariantType {
+	GDNATIVE_VARIANT_TYPE_NIL,
+
+	// atomic types
+	GDNATIVE_VARIANT_TYPE_BOOL,
+	GDNATIVE_VARIANT_TYPE_INT,
+	GDNATIVE_VARIANT_TYPE_FLOAT,
+	GDNATIVE_VARIANT_TYPE_STRING,
+
+	// math types
+	GDNATIVE_VARIANT_TYPE_VECTOR2,
+	GDNATIVE_VARIANT_TYPE_VECTOR2I,
+	GDNATIVE_VARIANT_TYPE_RECT2,
+	GDNATIVE_VARIANT_TYPE_RECT2I,
+	GDNATIVE_VARIANT_TYPE_VECTOR3,
+	GDNATIVE_VARIANT_TYPE_VECTOR3I,
+	GDNATIVE_VARIANT_TYPE_TRANSFORM2D,
+	GDNATIVE_VARIANT_TYPE_PLANE,
+	GDNATIVE_VARIANT_TYPE_QUATERNION,
+	GDNATIVE_VARIANT_TYPE_AABB,
+	GDNATIVE_VARIANT_TYPE_BASIS,
+	GDNATIVE_VARIANT_TYPE_TRANSFORM3D,
+
+	// misc types
+	GDNATIVE_VARIANT_TYPE_COLOR,
+	GDNATIVE_VARIANT_TYPE_STRING_NAME,
+	GDNATIVE_VARIANT_TYPE_NODE_PATH,
+	GDNATIVE_VARIANT_TYPE_RID,
+	GDNATIVE_VARIANT_TYPE_OBJECT,
+	GDNATIVE_VARIANT_TYPE_CALLABLE,
+	GDNATIVE_VARIANT_TYPE_SIGNAL,
+	GDNATIVE_VARIANT_TYPE_DICTIONARY,
+	GDNATIVE_VARIANT_TYPE_ARRAY,
+
+	// typed arrays
+	GDNATIVE_VARIANT_TYPE_PACKED_BYTE_ARRAY,
+	GDNATIVE_VARIANT_TYPE_PACKED_INT32_ARRAY,
+	GDNATIVE_VARIANT_TYPE_PACKED_INT64_ARRAY,
+	GDNATIVE_VARIANT_TYPE_PACKED_FLOAT32_ARRAY,
+	GDNATIVE_VARIANT_TYPE_PACKED_FLOAT64_ARRAY,
+	GDNATIVE_VARIANT_TYPE_PACKED_STRING_ARRAY,
+	GDNATIVE_VARIANT_TYPE_PACKED_VECTOR2_ARRAY,
+	GDNATIVE_VARIANT_TYPE_PACKED_VECTOR3_ARRAY,
+	GDNATIVE_VARIANT_TYPE_PACKED_COLOR_ARRAY,
+
+	GDNATIVE_VARIANT_TYPE_VARIANT_MAX
+};
+
+enum GDNativeVariantOperator {
+	//comparison
+	NATIVE_VARIANT_OP_EQUAL,
+	NATIVE_VARIANT_OP_NOT_EQUAL,
+	NATIVE_VARIANT_OP_LESS,
+	NATIVE_VARIANT_OP_LESS_EQUAL,
+	NATIVE_VARIANT_OP_GREATER,
+	NATIVE_VARIANT_OP_GREATER_EQUAL,
+	//mathematic
+	NATIVE_VARIANT_OP_ADD,
+	NATIVE_VARIANT_OP_SUBTRACT,
+	NATIVE_VARIANT_OP_MULTIPLY,
+	NATIVE_VARIANT_OP_DIVIDE,
+	NATIVE_VARIANT_OP_NEGATE,
+	NATIVE_VARIANT_OP_POSITIVE,
+	NATIVE_VARIANT_OP_MODULE,
+	//bitwise
+	NATIVE_VARIANT_OP_SHIFT_LEFT,
+	NATIVE_VARIANT_OP_SHIFT_RIGHT,
+	NATIVE_VARIANT_OP_BIT_AND,
+	NATIVE_VARIANT_OP_BIT_OR,
+	NATIVE_VARIANT_OP_BIT_XOR,
+	NATIVE_VARIANT_OP_BIT_NEGATE,
+	//logic
+	NATIVE_VARIANT_OP_AND,
+	NATIVE_VARIANT_OP_OR,
+	NATIVE_VARIANT_OP_XOR,
+	NATIVE_VARIANT_OP_NOT,
+	//containment
+	NATIVE_VARIANT_OP_IN,
+	NATIVE_VARIANT_OP_MAX
+
+};
+
+typedef void *GDNativeVariantPtr;
+typedef void *GDNativeStringNamePtr;
```

Comment:
This and other functions use `GDNativeStringNamePtr`, so we should add helpers to convert from `const char *` to `StringName` and vice versa.

Or maybe these should just use `const char *` and let Godot do the conversion internally?

Reply:
For this specific case, the only places where StringName is used is for the variant functions, for which I assume the binder has generated the right interfaces to use, so I am not sure its worth adding this. Generating a string from a CString and then a stringname from the right constructor is fast anyway.